keystore/v2/keystore/filesystem/backend/file_lock.go

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package backend
+
+import (
+	"os"
+	"sync"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// fileLock is an interprocess read-write lock that serializes access to filesystem.
+//
+// This lock is implemented with BSD flock(2) and has corresponding semantics.
+//
+// Note that this is an advisory lock. That is, this lock can be used to guard
+// access to some filesystem resource by cooperating processes, but it does not
+// guarantee correct serialization of filesystem accesses by itself.
+//
+// Also note that flock(2) has subtly different semantics from POSIX record locks
+// created by fcntl(2). Read corresponding system manual pages for details.
+type fileLock struct {
+	lockFile *os.File
+	path     string
+	// Here's a thing: BSD file locks are per-process. That is, they are shared
+	// between threads like fds. Furthermore, they can be 'transparently'
+	// upgraded: if the process holds a shared lock and grabs it again as exclusive,
+	// the system will upgrade the lock to exclusive (possibly by releasing and
+	// re-grabbing it). This is not what we want, so keep a regular lock here
+	// to make sure that goroutines are correctly synchronized among themselves.
+	lockSync sync.Mutex
+}
+
+// newFileLock creates a lock anchored at given file path.
+func newFileLock(path string) (*fileLock, error) {
+	lock, err := os.Create(path)
+	if err != nil {
+		return nil, err
+	}
+	return &fileLock{lockFile: lock, path: path}, nil
+}
+
+// Close the lock. This releases the lock if it is held.
+func (l *fileLock) Close() error {
+	return l.lockFile.Close()
+}
+
+// We need to juggle *two* locks here -- the mutex and the file lock -- and
+// this is hard. So hard that we are using a regular sync.Mutex here rather
+// than seemingly more natural sync.RWMutex. Lock contention within the
+// process should be fairly minimal, and different processes still use the
+// read-write semantics, so it's okay.
+//
+// The reason for this complexity is that unlocking the flock can fail and
```

Comment:
> did you meet situations when flock failed? when it can occur?

No, I did not encounter situations like this. The documentation for flock() does not mention obvious and practical ways to shoot yourself in the foot. However, I'd imagine that it can fail in situations like

  - `Close()` followed by `Unlock()` causes flock() to be called for invalid file descriptor;
  - if someone puts key store on sshfs or other networked filesystem and it fails;
  - if key store is located on some removable media which gets removed with lock held;
  - some other weird failure in the kernel because of kernel reasons.

So these are largely theoretic possibilities. However, we have a practical issue of syscall.Flock() returning an error that we need to handle in some way. I felt that just ignoring it, hoping for the best, and – most importantly – risking with a possible deadlock is not robust enough. I'd be maybe okay with ignoring an error that causes other things to fail loudly with screams in logs, but if that's going to be a deadlock that silently blocks the service and makes it unresponsive, then it's definitely not okay.